### DISCUSSION TITLE
Send "shutdown complete message" over UART?

### DISCUSSION THREAD

**Discussion by hongkongkiwi** [finit-project/finit#287] [Q&A] [answered]
In my board after everything is unmounted and it's safe to shutdown or reboot, I wanted to push a command over UART (so it needs to complete sending the command, but we don't need to wait for a response).

Essentially this is telling a slave MCU that it's safe to toggle the power off when in this state.

I was hoping to get your idea on the best way to achieve this in Finit? Should this be a plugin? or can it be a script somehow which is hooked at that exact state?

The idea is I really want to make sure this command successfully runs as the second to last action (and the command completes) then the system can shutdown or reboot.

I want all filesystems to unmount except root (as we are accessing /dev/ttyS2 to send the message).

**Comment by troglobit** [comment, marked as answer]
Interesting! When we designed a lot of the logic in Finit it was around making sure the system never shut down, but instead was kept alive and instead rebooted on error.  For this we did a very tight integration with [watchdogd](https://github.com/troglobit/watchdogd/), which I developed closely with Finit.  We have very clever integration of these two, in fact, Finit comes with its own finit-watchdogd which services can be handed over to watchdogd when it starts up.  As long as there's one of them available we let it do the actual reset logic -- by saving reset cause, setting the lowest WDT timeout in the kernel, and then allow the WDT to perform the board reset.  If this fails, for some reason, Finit falls back to call `reboot(RB_AUTOBOOT);`, letting the kernel perform the reboot.  This fallback was not guaranteed however to reset all logic, in particular many i2c I/O expanders were not reset properly, so we would end up in a bad state, having just the SoC reboot.  Later we tried to instead hook into the board reset code path in the kernel to handle this better.  Maybe that's something you could do as well?

So, that side of things is very developed, but the code is there in `src/sig.c:do_shutdown()`.  As you can see, at power-off Finit calls `reboot(RB_POWER_OFF);`, so one way to handle this is to simply patch Finit to do this, another is to perhaps look into if there is any pins on your SoC going low, or sth, the MCU can react to instead of using the serial port?

**Reply by troglobit** [reply]
In the latest batch of changes pushed to master, we now have two new hook points late in shutdown and support for running custom scripts:

https://github.com/troglobit/finit/blob/505c370898fe548ea43f6744272796796b739d25/src/sig.c#L380-L381

This one is right before the kernel syscall to reboot/poweroff the system.  For details on how to enable this optional feature, and add hook scripts to your system, see https://github.com/troglobit/finit/blob/master/doc/plugins.md#hooks

**Comment by hongkongkiwi** [comment]
I wanted to follow up on this.

I have a need to force a reboot using the watchdog or custom UART command to a slave SoC (because I need to toggle the reset pin of the Soc) whenever a soft reset is requested instead of using the built in kernel reset function.

What's the best way to do this? I could modify the kernel behaviour but that's a bit annoying.

I thought of using a hook script in finit at the last stage to send a UART message and request the slave MCU to reset.

Is there any issue with doing this as the last hook point?

Whats the best way to accomplish this?

**Reply by troglobit** [reply]
I would say the best way is to use a hook script at `HOOK_SYSTEM_DN`, as I mentioned in the comment here (above) https://github.com/troglobit/finit/discussions/287?notification_referrer_id=NT_kwDOAALM3bEzODEyMDk3NzA5OjE4MzUxNw#discussioncomment-4097402 :-)
